File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/library/items/media/files/uri/specs/GivenAFileThatIsStoredCachedAndAvailableRemotely/WhenGettingTheUri.java
package com.lasthopesoftware.bluewater.client.library.items.media.files.uri.specs.GivenAFileThatIsStoredCachedAndAvailableRemotely;

import android.net.Uri;

import com.lasthopesoftware.bluewater.client.library.items.media.files.ServiceFile;
import com.lasthopesoftware.bluewater.client.library.items.media.files.stored.IStoredFileAccess;
import com.lasthopesoftware.bluewater.client.library.items.media.files.stored.repository.StoredFile;
import com.lasthopesoftware.bluewater.client.library.items.media.files.stored.system.uri.MediaFileUriProvider;
import com.lasthopesoftware.bluewater.client.library.items.media.files.stored.uri.StoredFileUriProvider;
import com.lasthopesoftware.bluewater.client.library.items.media.files.uri.BestMatchUriProvider;
import com.lasthopesoftware.bluewater.client.library.items.media.files.uri.RemoteFileUriProvider;
import com.lasthopesoftware.bluewater.client.library.repository.Library;
import com.namehillsoftware.handoff.promises.Promise;

import org.junit.Test;

import java.util.concurrent.CountDownLatch;

import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class WhenGettingTheUri {

	private static Throwable rejection;
	private static Uri returnedFileUri;

	public static void context() throws Throwable {
		final IStoredFileAccess storedFileAccess = mock(IStoredFileAccess.class);
		when(storedFileAccess.getStoredFile(new ServiceFile(3))).thenReturn(
			new Promise<>(new StoredFile(
				new Library(),
				1,
				new ServiceFile(3),
				"file:///a_path",
				true)));

		final BestMatchUriProvider bestMatchUriProvider =
			new BestMatchUriProvider(
				new Library(),
				new StoredFileUriProvider(
					storedFileAccess,
					() -> true),
				mock(MediaFileUriProvider.class),
				mock(RemoteFileUriProvider.class));

		final CountDownLatch countDownLatch = new CountDownLatch(1);
		bestMatchUriProvider
			.getFileUri(new ServiceFile(3))
			.then(f -> {
				returnedFileUri = f;
				countDownLatch.countDown();
				return null;
			})
			.excuse(e -> {
				rejection = e;
				return null;
			});

		countDownLatch.await();

		if (rejection != null)
			throw rejection;
	}

	@Test
	public void thenTheStoredFileUriIsReturned() {
		assertThat(returnedFileUri.toString()).isEqualTo("file:///a_path");
	}
}

File: projectBlueWater/src/test/java/com/lasthopesoftware/bluewater/client/browsing/items/media/files/uri/GivenAFileThatIsStoredCachedAndAvailableRemotely/WhenGettingTheUri.kt
package com.lasthopesoftware.bluewater.client.browsing.items.media.files.uri.GivenAFileThatIsStoredCachedAndAvailableRemotely

import android.net.Uri
import com.lasthopesoftware.bluewater.client.browsing.items.media.audio.uri.CachedAudioFileUriProvider
import com.lasthopesoftware.bluewater.client.browsing.items.media.files.ServiceFile
import com.lasthopesoftware.bluewater.client.browsing.items.media.files.uri.BestMatchUriProvider
import com.lasthopesoftware.bluewater.client.browsing.items.media.files.uri.RemoteFileUriProvider
import com.lasthopesoftware.bluewater.client.browsing.library.repository.Library
import com.lasthopesoftware.bluewater.client.stored.library.items.files.system.uri.MediaFileUriProvider
import com.lasthopesoftware.bluewater.client.stored.library.items.files.uri.StoredFileUriProvider
import com.lasthopesoftware.bluewater.shared.promises.extensions.toFuture
import com.namehillsoftware.handoff.promises.Promise
import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.AssertionsForClassTypes.assertThat
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import java.io.File

@RunWith(RobolectricTestRunner::class)
class WhenGettingTheUri {

	companion object {
		private val returnedFileUri by lazy {
			val mockStoredFileUriProvider = mockk<StoredFileUriProvider>()
			every { mockStoredFileUriProvider.promiseFileUri(any()) } returns Promise(Uri.fromFile(File("/a_path/to_a_file.mp3")))

			val cachedAudioFileUriProvider = mockk<CachedAudioFileUriProvider>()
			every { cachedAudioFileUriProvider.promiseFileUri(ServiceFile(3)) } returns Promise(Uri.fromFile(File("/a_cached_path/to_a_file.mp3")))

			val mockMediaFileUriProvider = mockk<MediaFileUriProvider>()
			every { mockMediaFileUriProvider.promiseFileUri(any()) } returns Promise(Uri.fromFile(File("/a_media_path/to_a_file.mp3")))

			val mockRemoteFileUriProvider = mockk<RemoteFileUriProvider>()
			every { mockRemoteFileUriProvider.promiseFileUri(ServiceFile(3)) } returns Promise(Uri.parse("http://remote-url/to_a_file.mp3"))

			val bestMatchUriProvider = BestMatchUriProvider(
				Library(),
				mockStoredFileUriProvider,
				cachedAudioFileUriProvider,
				mockMediaFileUriProvider,
				mockRemoteFileUriProvider
			)
			bestMatchUriProvider
				.promiseFileUri(ServiceFile(3))
				.toFuture()
				.get()
		}
	}

	@Test
	fun thenTheStoredFileUriIsReturned() {
		assertThat(returnedFileUri.toString())
			.isEqualTo("file:///a_path/to_a_file.mp3")
	}
}
